orc8r/cloud/go/services/bootstrapper/servicers/registration/cloud_registration.go

```diff
@@ -0,0 +1,147 @@
+/*
+Copyright 2020 The Magma Authors.
+
+This source code is licensed under the BSD-style license found in the
+LICENSE file in the root directory of this source tree.
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package registration
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"magma/orc8r/cloud/go/clock"
+	"magma/orc8r/lib/go/protos"
+
+	"github.com/golang/glog"
+	"github.com/golang/protobuf/ptypes/timestamp"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+const (
+	// NonceLength is the number of characters that the nonce will have
+	NonceLength = 30
+
+	// NotImplementedWarning is pulled out of getDomainName for ease of testing
+	NotImplementedWarning = "warning: not implemented"
+)
+
+type cloudRegistrationServicer struct {
+	store   Store
+	rootCA  string
+	timeout time.Duration
+}
+
+func NewCloudRegistrationServicer(store Store, rootCA string, timeout time.Duration) (protos.CloudRegistrationServer, error) {
+	if store == nil {
+		return nil, fmt.Errorf("storage store is nil")
+	}
+	return &cloudRegistrationServicer{store: store, rootCA: rootCA, timeout: timeout}, nil
+}
+
+func (c *cloudRegistrationServicer) GetToken(ctx context.Context, request *protos.GetTokenRequest) (*protos.GetTokenResponse, error) {
+	networkId := request.GatewayDeviceInfo.NetworkId
+	logicalId := request.GatewayDeviceInfo.LogicalId
+
+	tokenInfo, err := c.store.GetTokenInfoFromLogicalID(networkId, logicalId)
+	if err != nil {
+		// Error is not bubbled up since the tokenInfo is only important if the token exists and is expired
+		// If GetTokenInfoFromLogicalID fails, we can just ignore and continue
+		glog.V(2).Infof("could not get tokenInfo for networkID %v and logicalID %v: %v", networkId, logicalId, err)
+	}
+
+	refresh := request.Refresh || tokenInfo == nil || IsExpired(tokenInfo)
```

Comment:
I feel like maybe we should remove refresh -- it adds extra complexity. If so, start the removal from the setup PR

Reply:
Will do if bopped, but won't if not